hw/drivers/uart/uart_hal/src/uart_hal.c

```diff
@@ -82,31 +104,70 @@ uart_hal_open(struct os_dev *odev, uint32_t wait, void *arg)
     if (odev->od_flags & OS_DEV_F_STATUS_OPEN) {
         return OS_EBUSY;
     }
-    rc = hal_uart_init_cbs(uart_hal_dev_get_id(dev), uc->uc_tx_char, uc->uc_tx_done,
-                           uc->uc_rx_char, uc->uc_cb_arg);
+
+    dev->ud_conf = *uc;
+
+    return uart_hal_configure(dev);
+}
+
+static int
+uart_hal_close(struct os_dev *odev)
+{
+    struct uart_dev *dev;
+    int rc;
+
+    dev = (struct uart_dev *)odev;
+    rc = hal_uart_close(uart_hal_dev_get_id(dev));
     if (rc) {
         return OS_EINVAL;
     }
 
-    rc = hal_uart_config(uart_hal_dev_get_id(dev), uc->uc_speed, uc->uc_databits,
-      uc->uc_stopbits, (enum hal_uart_parity)uc->uc_parity, (enum hal_uart_flow_ctl)uc->uc_flow_ctl);
+    return OS_OK;
+}
+
+static int
+uart_hal_suspend(struct os_dev *odev, os_time_t suspend_at, int force)
+{
+    struct uart_dev *dev = (struct uart_dev *)odev;
+    int rc;
+
+    /*
+     * XXX this should not be used as an example on how to add suspend/resume
+     * to os_dev since it's just provisional implementation to allow disable
+     * and enable UART to have some basic power saving. The whole suspend/resume
+     * framework probably should be revised and proper implementation added here
+     * then.
+     */
+
+    assert(dev);
```

Comment:
assert() is unnecessary here. The caller would've already come via os_dev_suspend()